service/matching/version_sets.go

```diff
@@ -186,22 +231,50 @@ func findVersion(data *persistence.VersioningData, buildID string) findVersionRe
 	}
 }
 
-func makeDefaultSet(data *persistence.VersioningData, setIx int) {
+func makeDefaultSet(data *persistence.VersioningData, setIx int, timestamp *clockpb.HybridLogicalClock) {
+	data.DefaultUpdateTimestamp = timestamp
 	if len(data.VersionSets) <= 1 {
 		return
 	}
-	// Move the set to the end and shift all the others down
-	moveMe := data.VersionSets[setIx]
-	copy(data.VersionSets[setIx:], data.VersionSets[setIx+1:])
-	data.VersionSets[len(data.VersionSets)-1] = moveMe
+	if setIx < len(data.VersionSets)-1 {
+		// Move the set to the end and shift all the others down
+		moveMe := data.VersionSets[setIx]
+		copy(data.VersionSets[setIx:], data.VersionSets[setIx+1:])
+		data.VersionSets[len(data.VersionSets)-1] = moveMe
+	}
 }
 
-func makeVersionInSetDefault(data *persistence.VersioningData, setIx, versionIx int) {
+func makeVersionInSetDefault(data *persistence.VersioningData, setIx, versionIx int, timestamp *clockpb.HybridLogicalClock) {
+	data.VersionSets[setIx].DefaultUpdateTimestamp = timestamp
 	setVersions := data.VersionSets[setIx].BuildIds
 	if len(setVersions) <= 1 {
 		return
 	}
-	moveMe := setVersions[versionIx]
-	copy(setVersions[versionIx:], setVersions[versionIx+1:])
-	setVersions[len(setVersions)-1] = moveMe
+	if versionIx < len(setVersions)-1 {
+		// Move the build ID to the end and shift all the others down
+		moveMe := setVersions[versionIx]
+		copy(setVersions[versionIx:], setVersions[versionIx+1:])
+		setVersions[len(setVersions)-1] = moveMe
+	}
+}
+
+func generateHLCTimestamp(clock clockpb.HybridLogicalClock) clockpb.HybridLogicalClock {
+	wallclock := time.Now().UnixMilli()
```

Comment:
we usually use a `TimeSource` (from common/clock) instead of raw `time.Now()`. unfortunately `TimeSource` isn't powerful enough to be a true fake time, but maybe we should still use it for consistency, and making it easier to improve it in the future.

Reply:
I'm refactoring this in my next PR, please review there.